### PR TITLE
fix(hive): fix table location not match partition location

### PR DESCRIPTION
**Pull request by sandflee**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

assume hive table location  is `/a.db/b.table`,   for partiton `p_date = '20220820'`,  partition location is  `/a.db/b.table/p_date=20220820` by default,  but user may change partition location to  `/test/p_date=20220820`, 
so we should get sql data locations from partition location not table location

Fixes #issue

### CODE REVIEW

File: src/query/storages/hive/src/hive_catalog.rs [hunk 1]
```diff
@@ -81,6 +82,42 @@ impl HiveCatalog {
         Ok(ThriftHiveMetastoreSyncClient::new(i_prot, o_prot))
     }
 
+    pub async fn get_partitions_async(
```
Comment (BohuTANG) [inline src/query/storages/hive/src/hive_catalog.rs:85]: `get_partitions ` name is better?

File: src/query/storages/hive/src/hive_catalog.rs [hunk 2]
```diff
@@ -81,6 +82,42 @@ impl HiveCatalog {
...
+        &self,
+        db: String,
+        table: String,
+        partition_names: Vec<String>,
+    ) -> Result<Vec<Partition>> {
+        let client = self.get_client()?;
+        tokio::task::spawn_blocking(move || {
+            Self::do_get_partitions(client, db, table, partition_names)
+        })
+        .await
```
Comment (BohuTANG) [inline src/query/storages/hive/src/hive_catalog.rs:95]: `        .await?`

Comment (BohuTANG) [inline src/query/storages/hive/src/hive_catalog.rs:95]:
```suggestion
        .await?
```

File: src/query/storages/hive/src/hive_catalog.rs [hunk 3]
```diff
@@ -81,6 +82,42 @@ impl HiveCatalog {
...
+        .unwrap()
```
Comment (BohuTANG) [inline src/query/storages/hive/src/hive_catalog.rs:96]:
```suggestion
.await?
```

Comment (sandflee) [reply src/query/storages/hive/src/hive_catalog.rs:96]: if using '?' ,   some errors:
```
the trait `From<JoinError>` is not implemented for `ErrorCode`
```

Comment (sandflee) [reply src/query/storages/hive/src/hive_catalog.rs:96]: is it nesseserry to  convert JoinError to errorCode?

Comment (BohuTANG) [reply src/query/storages/hive/src/hive_catalog.rs:96]: No need. 
Please keep it.

File: src/query/storages/hive/src/hive_catalog.rs [hunk 4]
```diff
@@ -81,6 +82,42 @@ impl HiveCatalog {
...
     }
 
+    pub async fn get_partitions_async(
+        &self,
+        db: String,
+        table: String,
+        partition_names: Vec<String>,
+    ) -> Result<Vec<Partition>> {
+        let client = self.get_client()?;
+        tokio::task::spawn_blocking(move || {
```
Comment (BohuTANG) [inline src/query/storages/hive/src/hive_catalog.rs:92]: Why we need `spawn_blocking` here?

Comment (sandflee) [reply src/query/storages/hive/src/hive_catalog.rs:92]: thrift operator is io-blocking ,may block tokio worker thread

**Comment by BohuTANG** [comment]
cc @drmingdrmer 

https://github.com/datafuselabs/databend/runs/8090276309?check_suite_focus=true#step:4:2180

```
failures:

---- grpc::metasrv_grpc_schema_api::test_meta_grpc_client_cluster stdout ----
thread 'grpc::metasrv_grpc_schema_api::test_meta_grpc_client_cluster' panicked at 'called `Result::unwrap()` on an `Err` value: join cluster via ["127.0.0.1:29141"] fail

Stack backtrace:
   0: anyhow::error::<impl core::convert::From<E> for anyhow::Error>::from
   1: <core::result::Result<T,F> as core::ops::try_trait::FromResidual<core::result::Result<core::convert::Infallible,E>>>::from_residual
   2: it::tests::service::start_metasrv_with_context::{{closure}}
   3: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
   4: it::tests::service::start_metasrv_cluster::{{closure}}
   5: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
   6: <it::tests::service::MetaSrvBuilder as common_meta_api::kv_api::ApiBuilder<alloc::sync::Arc<common_meta_grpc::grpc_client::ClientHandle>>>::build_cluster::{{closure}}
   7: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
   8: <core::pin::Pin<P> as core::future::future::Future>::poll
   9: common_meta_api::schema_api_test_suite::SchemaApiTestSuite::test_cluster::{{closure}}
  10: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  11: it::grpc::metasrv_grpc_schema_api::test_meta_grpc_client_cluster::{{closure}}
  12: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
  13: <tracing::instrument::Instrumented<T> as core::future::future::Future>::poll
  14: tokio::park::thread::CachedParkThread::block_on::{{closure}}
  15: tokio::coop::with_budget::{{closure}}
  16: std::thread::local::LocalKey<T>::try_with
  17: std::thread::local::LocalKey<T>::with
  18: tokio::park::thread::CachedParkThread::block_on
  19: tokio::runtime::enter::Enter::block_on
  20: tokio::runtime::thread_pool::ThreadPool::block_on
  21: tokio::runtime::Runtime::block_on
  22: it::grpc::metasrv_grpc_schema_api::test_meta_grpc_client_cluster
  23: it::grpc::metasrv_grpc_schema_api::test_meta_grpc_client_cluster::{{closure}}
  24: core::ops::function::FnOnce::call_once
  25: core::ops::function::FnOnce::call_once
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/core/src/ops/function.rs:248:5
  26: test::__rust_begin_short_backtrace
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/test/src/lib.rs:572:5
  27: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/alloc/src/boxed.rs:1951:9
  28: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/core/src/panic/unwind_safe.rs:271:9
  29: std::panicking::try::do_call
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/std/src/panicking.rs:492:40
  30: std::panicking::try
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/std/src/panicking.rs:456:19
  31: std::panic::catch_unwind
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/std/src/panic.rs:137:14
  32: test::run_test_in_process
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/test/src/lib.rs:595:18
  33: test::run_test::run_test_inner::{{closure}}
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/test/src/lib.rs:489:39
  34: test::run_test::run_test_inner::{{closure}}
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/test/src/lib.rs:516:37
  35: std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/std/src/sys_common/backtrace.rs:122:18
  36: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/std/src/thread/mod.rs:501:17
  37: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/core/src/panic/unwind_safe.rs:271:9
  38: std::panicking::try::do_call
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/std/src/panicking.rs:492:40
  39: std::panicking::try
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/std/src/panicking.rs:456:19
  40: std::panic::catch_unwind
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/std/src/panic.rs:137:14
  41: std::thread::Builder::spawn_unchecked_::{{closure}}
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/std/src/thread/mod.rs:500:30
  42: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/core/src/ops/function.rs:248:5
  43: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/alloc/src/boxed.rs:1951:9
  44: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/alloc/src/boxed.rs:1951:9
  45: std::sys::unix::thread::Thread::new::thread_start
             at /rustc/ddcbba036aee08f0709f98a92a342a278eae5c05/library/std/src/sys/unix/thread.rs:108:17
  46: start_thread
  47: clone', src/meta/service/tests/it/tests/service.rs:242:59
stack backtrace:
   0:     0x55596f6d627d - std::backtrace_rs::backtrace::libunwind::trace::h79327ccd22e2164a
```

**Comment by sandflee** [comment]
@mergify update

**Review by dantengsky — APPROVED** [review]
LGTM

**Comment by drmingdrmer** [comment]
> cc @drmingdrmer
> 
> https://github.com/datafuselabs/databend/runs/8090276309?check_suite_focus=true#step:4:2180
> 
> ```
> failures:
> 
> ---- grpc::metasrv_grpc_schema_api::test_meta_grpc_client_cluster stdout ----
> thread 'grpc::metasrv_grpc_schema_api::test_meta_grpc_client_cluster' panicked at 'called `Result::unwrap()` on an `Err` value: join cluster via ["127.0.0.1:29141"] fail
> ```

It timed out joining a testing cluster. Just ignore it for now.

**Review by BohuTANG — APPROVED** [review]

**Pull request merged**